### PR TITLE
AP_RangeFinder: added trone in-tree driver

### PR DESCRIPTION
**Pull request by tridge**
This is pending testing by Luis Rodrigues

### CODE REVIEW

File: libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+#include <AP_HAL/AP_HAL.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define TRONE_I2C_ADDR 0x30
+
+// registers
+#define TRONE_MEASURE 0x00
+#define TRONE_WHOAMI  0x01
+#define TRONE_WHOAMI_VALUE 0xA1
```
Comment (OXINARF) [inline libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp]: The WHOAMI definitions are unused, did you find it to be unreliable?

File: libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+/*
+  timer called at 20Hz
+*/
+bool AP_RangeFinder_trone::timer(void)
+{
+    // take a reading
+    uint16_t distance_cm;
+    if (collect(distance_cm) && _sem->take(0)) {
```
Comment (OXINARF) [inline libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp]: Should this really be 0? I mean, that's a block indefinitely that can happen in the air.

File: libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    return true;
+}
+
+
+/*
+   update the state of the sensor
+*/
+void AP_RangeFinder_trone::update(void)
+{
+    if (_sem->take(0)) {
```
Comment (OXINARF) [inline libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp]: Same here.

File: libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp (hunk 2 above)
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp]: Shouldn't the take be first?

**Comment by lfrodrigues** [comment]
all good :)

File: libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp (hunk 1 above)
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp]: they should be used, I'll fix. I don't have a trone to test with btw, I'm relying on Luis to test

File: libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp (hunk 2 above)
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp]: if should be take(0), yes. This is not a bus semaphore, its a data structure semaphore. It should never be used to protect any bus operations. So it is safe (and correct) to always use it with no timeout as all callers are only doing local calculations with the semaphore held.
The take is 2nd as we don't want to hold the semaphore while doing any bus operation. The bus semaphore is held before timer is called

**Comment by tridge** [comment]
@lfrodrigues, I'll add the WHOAMI check if you don't mind re-testing

File: libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp (hunk 2 above)
Comment (OXINARF) [inline libraries/AP_RangeFinder/AP_RangeFinder_trone.cpp]: I know this isn't a bus semaphore, but it just takes a missing *give()* to block it forever while in the air (and as this is in hardware drivers it isn't tested in SITL, so not that hard to miss a bug like that).

**Comment by tridge** [comment]
I've added the WHOAMI and CRC checks, @lfrodrigues, please re-test

**Comment by tridge** [comment]
applied

**Pull request closed without merging**